In [1]:
import pandas as pd
import numpy as np
import os
import warnings

# 1. Preprocessing

In [2]:
warnings.filterwarnings("ignore")

In [3]:
flight_dir_2016 = r'2016'
flight_dir_2017 = r'2017'

In [4]:
weather_dir = r"/Weather"

In [5]:
airport_code = ["ALT", "CLT", "DEN", "DFW", "EWR", "IAH", "JFK", "LAS" ,"LAX", "MCO",
               "MIA", "ORD", "PHX", "SEA", "SFO"]

In [6]:
# weather_cols = ['windspeedKmph', 'windirDegree', 'weatherCode', 'percipMM', 'visibilty', 'pressure', 'cloudcover', 'DewPointF', 'WindGustKmph',
#                 'tempF', 'WindChillF', 'humidity', 'date',  'time', 'airport']


In [7]:
flight_cols = ["FlightDate", "Quarter", "Year", "Month", "DayofMonth", "DepTime",
              "DepDel15", "CRSDepTime", "DepDelayMinutes", "OriginAirportID", "DestAirportID",
              "ArrTime", "CRSArrTime", "ArrDel15", "ArrDelayMinutes"]

## Flight data

In [8]:
Folders_2016 = os.listdir(flight_dir_2016)
Folders_2016.sort()
if '.DS_Store' in Folders_2016:
    Folders_2016.remove('.DS_Store')
Folders_2017 = os.listdir(flight_dir_2017)
Folders_2017.sort()
if '.DS_Store' in Folders_2017:
    Folders_2017.remove('.DS_Store')

In [9]:
print(Folders_2016)
print()
print(Folders_2017)

['On_Time_On_Time_Performance_2016_1', 'On_Time_On_Time_Performance_2016_10', 'On_Time_On_Time_Performance_2016_11', 'On_Time_On_Time_Performance_2016_12', 'On_Time_On_Time_Performance_2016_2', 'On_Time_On_Time_Performance_2016_3', 'On_Time_On_Time_Performance_2016_4', 'On_Time_On_Time_Performance_2016_5', 'On_Time_On_Time_Performance_2016_6', 'On_Time_On_Time_Performance_2016_7', 'On_Time_On_Time_Performance_2016_8', 'On_Time_On_Time_Performance_2016_9']

['On_Time_On_Time_Performance_2017_1', 'On_Time_On_Time_Performance_2017_10', 'On_Time_On_Time_Performance_2017_11', 'On_Time_On_Time_Performance_2017_12', 'On_Time_On_Time_Performance_2017_2', 'On_Time_On_Time_Performance_2017_3', 'On_Time_On_Time_Performance_2017_4', 'On_Time_On_Time_Performance_2017_5', 'On_Time_On_Time_Performance_2017_6', 'On_Time_On_Time_Performance_2017_7', 'On_Time_On_Time_Performance_2017_8', 'On_Time_On_Time_Performance_2017_9']


In [10]:
def extract_files(mainDir, Folder):
    files = []
    for i in Folder:
        k = os.listdir(os.path.join(mainDir, i))
        k.sort()
        files.append(k[0])
    files.sort()
    return files

In [45]:
files_2016 = extract_files(flight_dir_2016, Folders_2016)
# files_2016

In [44]:
files_2017 = extract_files(flight_dir_2017, Folders_2017)
# files_2017

Target Rows: 1877296, 15

In [13]:
all_files = []
for i, folder in enumerate(Folders_2016):
    all_files.append(os.path.join(flight_dir_2016, folder, files_2016[i]))
for i, folder in enumerate(Folders_2017):
    all_files.append(os.path.join(flight_dir_2017, folder, files_2017[i])) 
all_files

['2016/On_Time_On_Time_Performance_2016_1/On_Time_On_Time_Performance_2016_1.csv',
 '2016/On_Time_On_Time_Performance_2016_10/On_Time_On_Time_Performance_2016_10.csv',
 '2016/On_Time_On_Time_Performance_2016_11/On_Time_On_Time_Performance_2016_11.csv',
 '2016/On_Time_On_Time_Performance_2016_12/On_Time_On_Time_Performance_2016_12.csv',
 '2016/On_Time_On_Time_Performance_2016_2/On_Time_On_Time_Performance_2016_2.csv',
 '2016/On_Time_On_Time_Performance_2016_3/On_Time_On_Time_Performance_2016_3.csv',
 '2016/On_Time_On_Time_Performance_2016_4/On_Time_On_Time_Performance_2016_4.csv',
 '2016/On_Time_On_Time_Performance_2016_5/On_Time_On_Time_Performance_2016_5.csv',
 '2016/On_Time_On_Time_Performance_2016_6/On_Time_On_Time_Performance_2016_6.csv',
 '2016/On_Time_On_Time_Performance_2016_7/On_Time_On_Time_Performance_2016_7.csv',
 '2016/On_Time_On_Time_Performance_2016_8/On_Time_On_Time_Performance_2016_8.csv',
 '2016/On_Time_On_Time_Performance_2016_9/On_Time_On_Time_Performance_2016_9.csv'

In [14]:
def retrieve_data(files):
    global Main_df
    for file in files:
        data = pd.read_csv(file)
        
        data = data[flight_cols+['Origin', 'Dest']]
        
        data = data[(data['Origin'].isin(airport_code)) & (data['Dest'].isin(airport_code))]

        Main_df = pd.concat([Main_df, data])

In [15]:
Main_df = pd.DataFrame()
retrieve_data(all_files)

In [16]:
Main_df.shape

(1578832, 17)

In [17]:
Main_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1578832 entries, 57 to 458711
Data columns (total 17 columns):
 #   Column           Non-Null Count    Dtype  
---  ------           --------------    -----  
 0   FlightDate       1578832 non-null  object 
 1   Quarter          1578832 non-null  int64  
 2   Year             1578832 non-null  int64  
 3   Month            1578832 non-null  int64  
 4   DayofMonth       1578832 non-null  int64  
 5   DepTime          1560778 non-null  float64
 6   DepDel15         1560771 non-null  float64
 7   CRSDepTime       1578832 non-null  int64  
 8   DepDelayMinutes  1560771 non-null  float64
 9   OriginAirportID  1578832 non-null  int64  
 10  DestAirportID    1578832 non-null  int64  
 11  ArrTime          1559814 non-null  float64
 12  CRSArrTime       1578832 non-null  int64  
 13  ArrDel15         1556652 non-null  float64
 14  ArrDelayMinutes  1556652 non-null  float64
 15  Origin           1578832 non-null  object 
 16  Dest             1

In [18]:
Df = Main_df.dropna() ## Drop all null values
Df['FlightDate'] = pd.to_datetime(Df['FlightDate']) # Convert object to pandas date type
Df.shape

(1556652, 17)

In [92]:
(23+1) % 24

0

In [137]:
def process_for_flight(x):
    date = x['FlightDate']
    time = x['DepTime']
#     print(type(date))
    time = str(int(time))
    time = time.zfill(4)
    hr, mns = time[:2], time[2:]
    flag = 0
    
    if int(mns) > 30 or hr=='24':  # I could see certain departure time is given as 24:00
        flag = 1
        hr = (int(hr) + 1) % 24

    if flag == 1 and hr == 0:
        one_day = pd.Timedelta(days=1)
        date += one_day
    
    return date, str(hr).zfill(2)

In [138]:
k = Df[['FlightDate','DepTime']].apply(process_for_flight, axis=1)
date = k.apply(lambda x: x[0])
hour = k.apply(lambda x: x[1])

In [139]:
Df['RoundedFlightDate'] = date
Df['DepatHr'] = hour

In [140]:
Df['Delayed'] = np.where(Df['DepDelayMinutes'] > 0, 1, 0)

In [141]:
Df.head()

FlightDate  Quarter  Year  Month  DayofMonth  DepTime  DepDel15  \
57 2016-01-01        1  2016      1           1    741.0       0.0   
58 2016-01-02        1  2016      1           2    737.0       0.0   
59 2016-01-03        1  2016      1           3    743.0       0.0   
60 2016-01-04        1  2016      1           4    737.0       0.0   
61 2016-01-05        1  2016      1           5    708.0       0.0   

    CRSDepTime  DepDelayMinutes  OriginAirportID  DestAirportID  ArrTime  \
57         745              0.0            14747          12478   1610.0   
58         745              0.0            14747          12478   1613.0   
59         745              0.0            14747          12478   1547.0   
60         745              0.0            14747          12478   1551.0   
61         710              0.0            14747          12478   1524.0   

    CRSArrTime  ArrDel15  ArrDelayMinutes Origin Dest RoundedFlightDate  \
57        1602       0.0              8.0    SEA  JFK        2016-01-01   
58        1602       0.0             11.0    SEA  JFK        2016-01-02   
59        1602       0.0              0.0    SEA  JFK        2016-01-03   
60        1602       0.0              0.0    SEA  JFK        2016-01-04   
61        1527       0.0              0.0    SEA  JFK        2016-01-05   

   DepatHr  Delayed  
57      08        0  
58      08        0  
59      08        0  
60      08        0  
61      07        0

In [142]:
print(f"Number of flights delayed: {Df['Delayed'].value_counts()[1]}, not delayed: {Df['Delayed'].value_counts()[0]}")

Number of flights delayed: 613684, not delayed: 942965


## Weather data

In [143]:
import json
from pandas import json_normalize
import re

In [144]:
def weather_files(directory):
    reqd_files = []
    Folders = os.listdir(directory)
    if '.DS_Store' in Folders:
        Folders.remove('.DS_Store')
    s = r'201(6|7)-\d{1,2}.json'
    for folder in Folders:
        all_files = os.listdir(os.path.join(directory, folder))
        for file in all_files:
            if re.search(s, file):
                reqd_files.append(os.path.join(directory, folder,file))
    reqd_files.sort()
    return reqd_files

In [145]:
all_weather_files = weather_files('Weather')
# all_weather_files

In [146]:
len(all_weather_files), 15 * 24

(360, 360)

In [147]:
def  merge_all_airports(all_weather_files):
    global Df2
    keys = ['windspeedKmph','winddirDegree', 'weatherCode', 'precipMM',
            'visibility', 'pressure', 'cloudcover', 'DewPointF', 'WindGustKmph',
            'tempF', 'WindChillF', 'humidity', 'time']
    for file in all_weather_files:
        airport = file.split('/')[1]
        #print(airport)
        f = open(file)
        D = json.load(f)
        data = json_normalize(D['data']['weather'])[['hourly', 'date']]
        data['date'] = pd.to_datetime(data['date'])
        data['airport'] = airport
        data = data.explode('hourly').reset_index(drop=True)
        data[keys] = pd.json_normalize(data['hourly'])[keys]
        data.drop(columns=['hourly'], inplace=True)
        Df2 = pd.concat([Df2, data])

In [148]:
Df2 = pd.DataFrame()
merge_all_airports(all_weather_files)

In [149]:
Df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 263160 entries, 0 to 719
Data columns (total 15 columns):
 #   Column         Non-Null Count   Dtype         
---  ------         --------------   -----         
 0   date           263160 non-null  datetime64[ns]
 1   airport        263160 non-null  object        
 2   windspeedKmph  263160 non-null  object        
 3   winddirDegree  263160 non-null  object        
 4   weatherCode    263160 non-null  object        
 5   precipMM       263160 non-null  object        
 6   visibility     263160 non-null  object        
 7   pressure       263160 non-null  object        
 8   cloudcover     263160 non-null  object        
 9   DewPointF      263160 non-null  object        
 10  WindGustKmph   263160 non-null  object        
 11  tempF          263160 non-null  object        
 12  WindChillF     263160 non-null  object        
 13  humidity       263160 non-null  object        
 14  time           263160 non-null  object        
dtypes: 

#### To do
- Merge with flight data based on flight departure time and weather time on a particular date


## Merging Flight data with weather

* (np.unique(Df['FlightDate']) == my_range).all() --> Seems like all date is present.

In [150]:
Df.drop_duplicates(inplace=True)

In [151]:
Df_sort = Df.sort_values(['FlightDate', 'DepatHr'])
# Df_sort = Df_sort.set_index('FlightDate')

In [152]:
## Flight Data
Df_sort.head()

FlightDate  Quarter  Year  Month  DayofMonth  DepTime  DepDel15  \
1435 2016-01-01        1  2016      1           1      6.0       1.0   
1497 2016-01-01        1  2016      1           1      6.0       0.0   
4839 2016-01-01        1  2016      1           1     25.0       0.0   
8434 2016-01-01        1  2016      1           1   2335.0       0.0   
9989 2016-01-01        1  2016      1           1     29.0       0.0   

      CRSDepTime  DepDelayMinutes  OriginAirportID  DestAirportID  ArrTime  \
1435        2347             19.0            12889          13303    721.0   
1497        2359              7.0            12892          13930    559.0   
4839          30              0.0            12892          13303    810.0   
8434        2335              0.0            12889          12478    710.0   
9989          30              0.0            14771          13303    832.0   

      CRSArrTime  ArrDel15  ArrDelayMinutes Origin Dest RoundedFlightDate  \
1435         714       0.0              7.0    LAS  MIA        2016-01-01   
1497         603       0.0              0.0    LAX  ORD        2016-01-01   
4839         823       0.0              0.0    LAX  MIA        2016-01-01   
8434         730       0.0              0.0    LAS  JFK        2016-01-02   
9989         850       0.0              0.0    SFO  MIA        2016-01-01   

     DepatHr  Delayed  
1435      00        1  
1497      00        1  
4839      00        0  
8434      00        0  
9989      00        0

In [153]:
def process_for_weather(x):
    x = str(int(x))
    #print(x)
    x = x.zfill(4)
    hr = x[:2]
    return  hr

In [154]:
Df2['timeInHr'] = Df2['time'].apply(process_for_weather) 

In [155]:
## Weather Data
Df2.head()

date airport windspeedKmph winddirDegree weatherCode precipMM  \
0 2016-01-01     ATL            11           318         176      0.1   
1 2016-01-01     ATL            13           317         176      0.0   
2 2016-01-01     ATL            14           315         122      0.0   
3 2016-01-01     ATL            16           314         122      0.0   
4 2016-01-01     ATL            17           314         122      0.0   

  visibility pressure cloudcover DewPointF WindGustKmph tempF WindChillF  \
0         10     1023        100        47           17    49         46   
1         10     1023        100        44           22    46         42   
2         10     1023        100        41           26    43         38   
3         10     1023        100        38           30    40         33   
4         10     1023        100        38           30    40         33   

  humidity time timeInHr  
0       91    0       00  
1       92  100       01  
2       92  200       02  
3       93  300       03  
4       93  400       04

#### Merging based on 
- Flight date and Departure hour from Df
- data and timeInHr from Df2

In [156]:
Df_sort.shape, Df2.shape

((1556649, 20), (263160, 16))

#### Note:
* When using left join, there will be NaN values(Because of non-availability of Weather data) - Rows: 1556649
    - Two issues:
        1. Certain entries have departure time as  **(24 + 1)%24 = 1**, which is wrong. We should have 00:00 next day date
            - Solution: Should manually deal with if,else condition
            - Around **291 rows**
        2. Entries that have 31/12/2017 and time more than 11:30  is **rounded to 01/01/2018** for which no weather data is present.
            - Solution: Either I have to retrieve weather data of January month or should Discard it.
            - **Only 8 rows** are present
* When using inner join, there will be no NaN values(Because only intersection is taken) - Rows: 1556350

In [166]:
Df3_left = pd.merge(Df_sort, Df2, left_on=['RoundedFlightDate', 'DepatHr', 'Origin'], right_on=['date', 'timeInHr', 'airport'], how='left' )
Df3_left.shape

(1556649, 36)

In [167]:
Df3 = pd.merge(Df_sort, Df2, left_on=['RoundedFlightDate', 'DepatHr', 'Origin'], right_on=['date', 'timeInHr', 'airport'], how='inner' )
Df3.shape

(1556641, 36)

In [168]:
Df3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1556641 entries, 0 to 1556640
Data columns (total 36 columns):
 #   Column             Non-Null Count    Dtype         
---  ------             --------------    -----         
 0   FlightDate         1556641 non-null  datetime64[ns]
 1   Quarter            1556641 non-null  int64         
 2   Year               1556641 non-null  int64         
 3   Month              1556641 non-null  int64         
 4   DayofMonth         1556641 non-null  int64         
 5   DepTime            1556641 non-null  float64       
 6   DepDel15           1556641 non-null  float64       
 7   CRSDepTime         1556641 non-null  int64         
 8   DepDelayMinutes    1556641 non-null  float64       
 9   OriginAirportID    1556641 non-null  int64         
 10  DestAirportID      1556641 non-null  int64         
 11  ArrTime            1556641 non-null  float64       
 12  CRSArrTime         1556641 non-null  int64         
 13  ArrDel15           1556641 

In [169]:
Df3.tail()

FlightDate  Quarter  Year  Month  DayofMonth  DepTime  DepDel15  \
1556636 2017-12-31        4  2017     12          31   2250.0       0.0   
1556637 2017-12-31        4  2017     12          31   2241.0       0.0   
1556638 2017-12-31        4  2017     12          31   2233.0       0.0   
1556639 2017-12-31        4  2017     12          31   2245.0       0.0   
1556640 2017-12-31        4  2017     12          31   2320.0       0.0   

         CRSDepTime  DepDelayMinutes  OriginAirportID  ...  visibility  \
1556636        2251              0.0            11292  ...          10   
1556637        2250              0.0            14771  ...          10   
1556638        2235              0.0            14771  ...          10   
1556639        2245              0.0            14771  ...          10   
1556640        2320              0.0            14771  ...          10   

         pressure  cloudcover  DewPointF  WindGustKmph tempF WindChillF  \
1556636      1038           7          2            13     5         -5   
1556637      1020          79         51             4    52         52   
1556638      1020          79         51             4    52         52   
1556639      1020          79         51             4    52         52   
1556640      1020          79         51             4    52         52   

        humidity  time  timeInHr  
1556636       83  2300        23  
1556637       97  2300        23  
1556638       97  2300        23  
1556639       97  2300        23  
1556640       97  2300        23  

[5 rows x 36 columns]

In [170]:
len(Df3_left[Df3_left.isna().any(axis=1)])

8

# 2. Classification of whether flight is delayed or not

## Pick only necessary columns

In [219]:
necessary_col = ['Quarter', 'Month',
        'Origin',
       'Dest', 'windspeedKmph', 'winddirDegree', 'weatherCode', 'precipMM',
       'visibility', 'pressure', 'cloudcover', 'DewPointF', 'WindGustKmph',
       'tempF', 'WindChillF', 'humidity', 'Delayed']

* Df3 - Complete merged dataframe
* Df4 - Datafame with just required columns for model creation

In [220]:
Df4 = Df3[necessary_col]

#### Either use
1. Ensemble model or
2. Neural Network(Pytorch)
    * But XGBoost is much better with tabular data

In [221]:
Df4.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1556641 entries, 0 to 1556640
Data columns (total 17 columns):
 #   Column         Non-Null Count    Dtype 
---  ------         --------------    ----- 
 0   Quarter        1556641 non-null  int64 
 1   Month          1556641 non-null  int64 
 2   Origin         1556641 non-null  object
 3   Dest           1556641 non-null  object
 4   windspeedKmph  1556641 non-null  object
 5   winddirDegree  1556641 non-null  object
 6   weatherCode    1556641 non-null  object
 7   precipMM       1556641 non-null  object
 8   visibility     1556641 non-null  object
 9   pressure       1556641 non-null  object
 10  cloudcover     1556641 non-null  object
 11  DewPointF      1556641 non-null  object
 12  WindGustKmph   1556641 non-null  object
 13  tempF          1556641 non-null  object
 14  WindChillF     1556641 non-null  object
 15  humidity       1556641 non-null  object
 16  Delayed        1556641 non-null  int64 
dtypes: int64(3), object(14)
mem

#### To Do:
1. Convert all data to numerical
1. **Import** appropriate model
1. Perform **cross validation** to find appropriate hyper parameteres, either with
    1. GridSearchCV or KFoldStratified
1. Determine model **performance**

In [222]:
Df4.columns

Index(['Quarter', 'Month', 'Origin', 'Dest', 'windspeedKmph', 'winddirDegree',
       'weatherCode', 'precipMM', 'visibility', 'pressure', 'cloudcover',
       'DewPointF', 'WindGustKmph', 'tempF', 'WindChillF', 'humidity',
       'Delayed'],
      dtype='object')

In [223]:
len(Df4['weatherCode'].value_counts())

43

In [253]:
Df4.iloc[:, 5:17] = Df4.iloc[:, 5:17].astype('float32')
Df4['weatherCode'] = Df4['weatherCode'].astype('object')
categorical = ['Quarter', 'Month', 'Origin', 'Dest', 'weatherCode']
numerical = [ 'windspeedKmph', 'winddirDegree', 'precipMM',
       'visibility', 'pressure', 'cloudcover', 'DewPointF', 'WindGustKmph',
       'tempF', 'WindChillF', 'humidity'] # Weather code actually can be categorical, but since we have 43 such categories it will be quite huge column set.

In [254]:
Df4.head()

Quarter  Month Origin Dest windspeedKmph  winddirDegree weatherCode  \
0        1      1    LAS  MIA            15          123.0       113.0   
1        1      1    LAX  ORD            15           38.0       113.0   
2        1      1    LAX  MIA            15           38.0       113.0   
3        1      1    LAX  DFW            15           38.0       113.0   
4        1      1    LAX  ORD            15           38.0       113.0   

   precipMM  visibility  pressure  cloudcover  DewPointF  WindGustKmph  tempF  \
0       0.0        10.0    1028.0         0.0       17.0          26.0   29.0   
1       0.0        10.0    1020.0         0.0       29.0          17.0   59.0   
2       0.0        10.0    1020.0         0.0       29.0          17.0   59.0   
3       0.0        10.0    1020.0         0.0       29.0          17.0   59.0   
4       0.0        10.0    1020.0         0.0       29.0          17.0   59.0   

   WindChillF  humidity  Delayed  
0        21.0      59.0      1.0  
1        58.0      39.0      1.0  
2        58.0      39.0      0.0  
3        58.0      39.0      1.0  
4        58.0      39.0      1.0

In [255]:
from xgboost import XGBClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import RandomizedSearchCV

Pipeline imports

In [256]:
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.compose import ColumnTransformer

In [257]:
categorical_pipeline = Pipeline(
    steps=[
        ("oh-encode", OneHotEncoder(handle_unknown="ignore", sparse=False)),
    ]
)
numerical_pipeline = Pipeline(
    steps=[
        ("scale", StandardScaler())
    ]
)

full_pipeline = ColumnTransformer(
    transformers=[
        ("numeric", numerical_pipeline, numerical),
        ("categorical", categorical_pipeline, categorical),
    ]
)

In [258]:
Df4.iloc[:,:-1].values

array([[1, 1, 'LAS', ..., 29.0, 21.0, 59.0],
       [1, 1, 'LAX', ..., 59.0, 58.0, 39.0],
       [1, 1, 'LAX', ..., 59.0, 58.0, 39.0],
       ...,
       [4, 12, 'SFO', ..., 52.0, 52.0, 97.0],
       [4, 12, 'SFO', ..., 52.0, 52.0, 97.0],
       [4, 12, 'SFO', ..., 52.0, 52.0, 97.0]], dtype=object)

In [260]:
X = full_pipeline.fit_transform(Df4.iloc[:,:-1])
X.shape

(1556641, 98)

In [262]:
y = Df4.iloc[:,-1]
y.shape

(1556641,)

##### Splitting
- But should not split after standardising because test data is standardised using the parameters of entire population, which is incorrect

In [265]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.8, stratify=y, random_state=2017)

In [266]:
X_train.shape, X_test.shape

((1245312, 98), (311329, 98))

In [268]:
xgb = XGBClassifier()

In [269]:
model = xgb.fit(X_train, y_train)

In [271]:
prediction = model.predict(X_test)

##### Evaluation
- Base model

In [272]:
from sklearn.metrics import accuracy_score, f1_score

In [273]:
accuracy_score(prediction, y_test )

0.6428312171368552

In [274]:
f1_score(prediction, y_test)

0.3838374882942588

# 3. Regression for Arrival delay period for Delayed flights